Copyright (c) Microsoft Corporation. All rights reserved.

![Impressions](https://PixelServer20190423114238.azurewebsites.net/api/impressions/NotebookVM/tutorials/regression-part2-automated-ml.png)

# Tutorial: Use automated machine learning to predict taxi fares

In this tutorial, you use automated machine learning in Azure Machine Learning service to create a regression model to predict NYC taxi fare prices. This process accepts training data and configuration settings, and automatically iterates through combinations of different feature normalization/standardization methods, models, and hyperparameter settings to arrive at the best model.

In this tutorial you learn the following tasks:

* Download, transform, and clean data using Azure Open Datasets
* Train an automated machine learning regression model
* Calculate model accuracy

If you don’t have an Azure subscription, create a free account before you begin. Try the [free or paid version](https://aka.ms/AMLFree) of Azure Machine Learning service today.

## Prerequisites

* Complete the [setup tutorial](https://docs.microsoft.com/azure/machine-learning/service/tutorial-1st-experiment-sdk-setup) if you don't already have an Azure Machine Learning service workspace or notebook virtual machine.
* After you complete the setup tutorial, open the **tutorials/regression-automated-ml.ipynb** notebook using the same notebook server.

This tutorial is also available on [GitHub](https://github.com/Azure/MachineLearningNotebooks/tree/master/tutorials) if you wish to run it in your own [local environment](https://github.com/Azure/MachineLearningNotebooks/blob/master/how-to-use-azureml/automated-machine-learning/README.md#setup-using-a-local-conda-environment).

## Download and prepare data

Import the necessary packages. The Open Datasets package contains a class representing each data source (`NycTlcGreen` for example) to easily filter date parameters before downloading.

In [1]:
import pandas as pd
from azureml.core import Dataset
from datetime import datetime
from dateutil.relativedelta import relativedelta

Begin by creating a dataframe to hold the taxi data. Then preview the data.

In [2]:
green_taxi_dataset = Dataset.Tabular.from_parquet_files(path="https://automlsamplenotebookdata.blob.core.windows.net/automl-sample-notebook-data/green_taxi_data.parquet")
green_taxi_df = green_taxi_dataset.to_pandas_dataframe()
green_taxi_df.head(10)

vendorID         lpepPickupDatetime        lpepDropoffDatetime  \
0         2 1970-01-01 00:23:42.643089 1970-01-01 00:23:42.644509   
1         1 1970-01-01 00:23:41.536899 1970-01-01 00:23:41.537716   
2         2 1970-01-01 00:23:41.372320 1970-01-01 00:23:41.373175   
3         2 1970-01-01 00:23:40.391366 1970-01-01 00:23:40.391772   
4         1 1970-01-01 00:23:41.230257 1970-01-01 00:23:41.231610   
5         2 1970-01-01 00:23:41.691041 1970-01-01 00:23:41.692340   
6         2 1970-01-01 00:23:40.127061 1970-01-01 00:23:40.127416   
7         2 1970-01-01 00:23:41.049681 1970-01-01 00:23:41.050492   
8         1 1970-01-01 00:23:41.445266 1970-01-01 00:23:41.446359   
9         2 1970-01-01 00:23:40.526093 1970-01-01 00:23:40.526663   

   passengerCount  tripDistance puLocationId doLocationId  pickupLongitude  \
0               1          1.88         None         None       -73.996155   
1               1          2.70         None         None       -73.978508   
2               1          3.54         None         None       -73.957787   
3               1          1.00         None         None       -73.919914   
4               1          5.10         None         None       -73.943710   
5               1          7.41         None         None       -73.940918   
6               1          1.03         None         None       -73.985718   
7               5          2.94         None         None       -73.939865   
8               1          3.00         None         None       -73.957939   
9               1          2.31         None         None       -73.943825   

   pickupLatitude  dropoffLongitude  ...  fareAmount  extra mtaTax  \
0       40.690903        -73.964287  ...        15.0    1.0    0.5   
1       40.687984        -73.955116  ...        11.5    0.5    0.5   
2       40.721779        -73.963005  ...        13.5    0.5    0.5   
3       40.826023        -73.904839  ...         6.5    0.0    0.5   
4       40.825439        -73.982964  ...        18.5    0.0    0.5   
5       40.839714        -73.994339  ...        24.0    0.0    0.5   
6       40.685646        -73.996773  ...         6.5    0.0    0.5   
7       40.789822        -73.952957  ...        12.5    0.0    0.5   
8       40.721928        -73.926247  ...        14.0    0.5    0.5   
9       40.810257        -73.943062  ...        10.0    0.0    0.5   

   improvementSurcharge  tipAmount  tollsAmount  ehailFee totalAmount  \
0                   0.3       4.00          0.0      None       20.80   
1                   0.3       2.55          0.0      None       15.35   
2                   0.3       2.80          0.0      None       17.60   
3                   0.3       0.00          0.0      None        7.30   
4                   0.3       3.85          0.0      None       23.15   
5                   0.3       4.80          0.0      None       29.60   
6                   0.3       1.30          0.0      None        8.60   
7                   0.3       0.00          0.0      None       13.30   
8                   0.3       2.00          0.0      None       17.30   
9                   0.3       2.00          0.0      None       12.80   

   tripType          __index_level_0__  
0       1.0 1970-01-01 00:23:42.643089  
1       1.0 1970-01-01 00:23:41.536899  
2       1.0 1970-01-01 00:23:41.372320  
3       1.0 1970-01-01 00:23:40.391366  
4       1.0 1970-01-01 00:23:41.230257  
5       1.0 1970-01-01 00:23:41.691041  
6       1.0 1970-01-01 00:23:40.127061  
7       1.0 1970-01-01 00:23:41.049681  
8       1.0 1970-01-01 00:23:41.445266  
9       1.0 1970-01-01 00:23:40.526093  

[10 rows x 24 columns]

Now that the initial data is loaded, define a function to create various time-based features from the pickup datetime field. This will create new fields for the month number, day of month, day of week, and hour of day, and will allow the model to factor in time-based seasonality. 

Use the `apply()` function on the dataframe to iteratively apply the `build_time_features()` function to each row in the taxi data.

In [3]:
def build_time_features(vector):
    pickup_datetime = vector[0]
    month_num = pickup_datetime.month
    day_of_month = pickup_datetime.day
    day_of_week = pickup_datetime.weekday()
    hour_of_day = pickup_datetime.hour
    
    return pd.Series((month_num, day_of_month, day_of_week, hour_of_day))

green_taxi_df[["month_num", "day_of_month","day_of_week", "hour_of_day"]] = green_taxi_df[["lpepPickupDatetime"]].apply(build_time_features, axis=1)
green_taxi_df.head(10)

vendorID         lpepPickupDatetime        lpepDropoffDatetime  \
0         2 1970-01-01 00:23:42.643089 1970-01-01 00:23:42.644509   
1         1 1970-01-01 00:23:41.536899 1970-01-01 00:23:41.537716   
2         2 1970-01-01 00:23:41.372320 1970-01-01 00:23:41.373175   
3         2 1970-01-01 00:23:40.391366 1970-01-01 00:23:40.391772   
4         1 1970-01-01 00:23:41.230257 1970-01-01 00:23:41.231610   
5         2 1970-01-01 00:23:41.691041 1970-01-01 00:23:41.692340   
6         2 1970-01-01 00:23:40.127061 1970-01-01 00:23:40.127416   
7         2 1970-01-01 00:23:41.049681 1970-01-01 00:23:41.050492   
8         1 1970-01-01 00:23:41.445266 1970-01-01 00:23:41.446359   
9         2 1970-01-01 00:23:40.526093 1970-01-01 00:23:40.526663   

   passengerCount  tripDistance puLocationId doLocationId  pickupLongitude  \
0               1          1.88         None         None       -73.996155   
1               1          2.70         None         None       -73.978508   
2               1          3.54         None         None       -73.957787   
3               1          1.00         None         None       -73.919914   
4               1          5.10         None         None       -73.943710   
5               1          7.41         None         None       -73.940918   
6               1          1.03         None         None       -73.985718   
7               5          2.94         None         None       -73.939865   
8               1          3.00         None         None       -73.957939   
9               1          2.31         None         None       -73.943825   

   pickupLatitude  dropoffLongitude  ...  tipAmount  tollsAmount ehailFee  \
0       40.690903        -73.964287  ...       4.00          0.0     None   
1       40.687984        -73.955116  ...       2.55          0.0     None   
2       40.721779        -73.963005  ...       2.80          0.0     None   
3       40.826023        -73.904839  ...       0.00          0.0     None   
4       40.825439        -73.982964  ...       3.85          0.0     None   
5       40.839714        -73.994339  ...       4.80          0.0     None   
6       40.685646        -73.996773  ...       1.30          0.0     None   
7       40.789822        -73.952957  ...       0.00          0.0     None   
8       40.721928        -73.926247  ...       2.00          0.0     None   
9       40.810257        -73.943062  ...       2.00          0.0     None   

   totalAmount  tripType          __index_level_0__  month_num day_of_month  \
0        20.80       1.0 1970-01-01 00:23:42.643089          1            1   
1        15.35       1.0 1970-01-01 00:23:41.536899          1            1   
2        17.60       1.0 1970-01-01 00:23:41.372320          1            1   
3         7.30       1.0 1970-01-01 00:23:40.391366          1            1   
4        23.15       1.0 1970-01-01 00:23:41.230257          1            1   
5        29.60       1.0 1970-01-01 00:23:41.691041          1            1   
6         8.60       1.0 1970-01-01 00:23:40.127061          1            1   
7        13.30       1.0 1970-01-01 00:23:41.049681          1            1   
8        17.30       1.0 1970-01-01 00:23:41.445266          1            1   
9        12.80       1.0 1970-01-01 00:23:40.526093          1            1   

   day_of_week  hour_of_day  
0            3            0  
1            3            0  
2            3            0  
3            3            0  
4            3            0  
5            3            0  
6            3            0  
7            3            0  
8            3            0  
9            3            0  

[10 rows x 28 columns]

Remove some of the columns that you won't need for training or additional feature building.

In [4]:
columns_to_remove = ["lpepPickupDatetime", "lpepDropoffDatetime", "puLocationId", "doLocationId", "extra", "mtaTax",
                     "improvementSurcharge", "tollsAmount", "ehailFee", "tripType", "rateCodeID", 
                     "storeAndFwdFlag", "paymentType", "fareAmount", "tipAmount"
                    ]
for col in columns_to_remove:
    green_taxi_df.pop(col)
    
green_taxi_df.head(5)

vendorID  passengerCount  tripDistance  pickupLongitude  pickupLatitude  \
0         2               1          1.88       -73.996155       40.690903   
1         1               1          2.70       -73.978508       40.687984   
2         2               1          3.54       -73.957787       40.721779   
3         2               1          1.00       -73.919914       40.826023   
4         1               1          5.10       -73.943710       40.825439   

   dropoffLongitude  dropoffLatitude  totalAmount          __index_level_0__  \
0        -73.964287        40.679707        20.80 1970-01-01 00:23:42.643089   
1        -73.955116        40.708138        15.35 1970-01-01 00:23:41.536899   
2        -73.963005        40.682774        17.60 1970-01-01 00:23:41.372320   
3        -73.904839        40.821404         7.30 1970-01-01 00:23:40.391366   
4        -73.982964        40.767857        23.15 1970-01-01 00:23:41.230257   

   month_num  day_of_month  day_of_week  hour_of_day  
0          1             1            3            0  
1          1             1            3            0  
2          1             1            3            0  
3          1             1            3            0  
4          1             1            3            0

### Cleanse data 

Run the `describe()` function on the new dataframe to see summary statistics for each field.

In [5]:
green_taxi_df.describe()

vendorID  passengerCount  tripDistance  pickupLongitude  \
count  24000.000000    24000.000000  24000.000000     24000.000000   
mean       1.777625        1.373625      2.893981       -73.827403   
std        0.415850        1.046180      3.072343         2.821767   
min        1.000000        0.000000      0.000000       -74.357101   
25%        2.000000        1.000000      1.050000       -73.959175   
50%        2.000000        1.000000      1.930000       -73.945049   
75%        2.000000        1.000000      3.700000       -73.917089   
max        2.000000        8.000000    154.280000         0.000000   

       pickupLatitude  dropoffLongitude  dropoffLatitude   totalAmount  \
count    24000.000000      24000.000000     24000.000000  24000.000000   
mean        40.689730        -73.819670        40.684436     14.892744   
std          1.556082          2.901199         1.599776     12.339749   
min          0.000000        -74.342766         0.000000   -120.800000   
25%         40.699127        -73.966476        40.699459      8.000000   
50%         40.746754        -73.944221        40.747536     11.300000   
75%         40.803060        -73.909061        40.791526     17.800000   
max         41.109089          0.000000        40.982826    425.000000   

       month_num  day_of_month  day_of_week  hour_of_day  
count    24000.0       24000.0      24000.0      24000.0  
mean         1.0           1.0          3.0          0.0  
std          0.0           0.0          0.0          0.0  
min          1.0           1.0          3.0          0.0  
25%          1.0           1.0          3.0          0.0  
50%          1.0           1.0          3.0          0.0  
75%          1.0           1.0          3.0          0.0  
max          1.0           1.0          3.0          0.0

From the summary statistics, you see that there are several fields that have outliers or values that will reduce model accuracy. First filter the lat/long fields to be within the bounds of the Manhattan area. This will filter out longer taxi trips or trips that are outliers in respect to their relationship with other features. 

Additionally filter the `tripDistance` field to be greater than zero but less than 31 miles (the haversine distance between the two lat/long pairs). This eliminates long outlier trips that have inconsistent trip cost.

Lastly, the `totalAmount` field has negative values for the taxi fares, which don't make sense in the context of our model, and the `passengerCount` field has bad data with the minimum values being zero.

Filter out these anomalies using query functions, and then remove the last few columns unnecessary for training.

In [6]:
final_df = green_taxi_df.query("pickupLatitude>=40.53 and pickupLatitude<=40.88")
final_df = final_df.query("pickupLongitude>=-74.09 and pickupLongitude<=-73.72")
final_df = final_df.query("tripDistance>=0.25 and tripDistance<31")
final_df = final_df.query("passengerCount>0 and totalAmount>0")

columns_to_remove_for_training = ["pickupLongitude", "pickupLatitude", "dropoffLongitude", "dropoffLatitude"]
for col in columns_to_remove_for_training:
    final_df.pop(col)

Call `describe()` again on the data to ensure cleansing worked as expected. You now have a prepared and cleansed set of taxi, holiday, and weather data to use for machine learning model training.

In [7]:
final_df.describe()

vendorID  passengerCount  tripDistance   totalAmount  month_num  \
count  23222.000000    23222.000000  23222.000000  23222.000000    23222.0   
mean       1.778572        1.374688      2.956753     14.838994        1.0   
std        0.415217        1.046995      2.862415     10.363600        0.0   
min        1.000000        1.000000      0.250000      0.010000        1.0   
25%        2.000000        1.000000      1.100000      8.190000        1.0   
50%        2.000000        1.000000      2.000000     11.750000        1.0   
75%        2.000000        1.000000      3.760000     17.880000        1.0   
max        2.000000        8.000000     30.840000    191.700000        1.0   

       day_of_month  day_of_week  hour_of_day  
count       23222.0      23222.0      23222.0  
mean            1.0          3.0          0.0  
std             0.0          0.0          0.0  
min             1.0          3.0          0.0  
25%             1.0          3.0          0.0  
50%             1.0          3.0          0.0  
75%             1.0          3.0          0.0  
max             1.0          3.0          0.0

## Configure workspace


Create a workspace object from the existing workspace. A [Workspace](https://docs.microsoft.com/python/api/azureml-core/azureml.core.workspace.workspace?view=azure-ml-py) is a class that accepts your Azure subscription and resource information. It also creates a cloud resource to monitor and track your model runs. `Workspace.from_config()` reads the file **config.json** and loads the authentication details into an object named `ws`. `ws` is used throughout the rest of the code in this tutorial.

In [8]:
from azureml.core.workspace import Workspace
ws = Workspace.from_config()

## Split the data into train and test sets

Split the data into training and test sets by using the `train_test_split` function in the `scikit-learn` library. This function segregates the data into the x (**features**) data set for model training and the y (**values to predict**) data set for testing. The `test_size` parameter determines the percentage of data to allocate to testing. The `random_state` parameter sets a seed to the random generator, so that your train-test splits are deterministic.

In [9]:
from sklearn.model_selection import train_test_split

x_train, x_test = train_test_split(final_df, test_size=0.2, random_state=223)

The purpose of this step is to have data points to test the finished model that haven't been used to train the model, in order to measure true accuracy. 

In other words, a well-trained model should be able to accurately make predictions from data it hasn't already seen. You now have data prepared for auto-training a machine learning model.

## Automatically train a model

To automatically train a model, take the following steps:
1. Define settings for the experiment run. Attach your training data to the configuration, and modify settings that control the training process.
1. Submit the experiment for model tuning. After submitting the experiment, the process iterates through different machine learning algorithms and hyperparameter settings, adhering to your defined constraints. It chooses the best-fit model by optimizing an accuracy metric.

### Define training settings

Define the experiment parameter and model settings for training. View the full list of [settings](https://docs.microsoft.com/azure/machine-learning/service/how-to-configure-auto-train). Submitting the experiment with these default settings will take approximately 20 minutes, but if you want a shorter run time, reduce the `experiment_timeout_hours` parameter.


|Property| Value in this tutorial |Description|
|----|----|---|
|**iteration_timeout_minutes**|10|Time limit in minutes for each iteration. Increase this value for larger datasets that need more time for each iteration.|
|**experiment_timeout_hours**|0.3|Maximum amount of time in hours that all iterations combined can take before the experiment terminates.|
|**enable_early_stopping**|True|Flag to enable early termination if the score is not improving in the short term.|
|**primary_metric**| spearman_correlation | Metric that you want to optimize. The best-fit model will be chosen based on this metric.|
|**featurization**| auto | By using auto, the experiment can preprocess the input data (handling missing data, converting text to numeric, etc.)|
|**verbosity**| logging.INFO | Controls the level of logging.|
|**n_cross_validations**|5|Number of cross-validation splits to perform when validation data is not specified.|

In [10]:
import logging

automl_settings = {
    "iteration_timeout_minutes": 10,
    "experiment_timeout_hours": 0.3,
    "enable_early_stopping": True,
    "primary_metric": 'spearman_correlation',
    "featurization": 'auto',
    "verbosity": logging.INFO,
    "n_cross_validations": 5
}

Use your defined training settings as a `**kwargs` parameter to an `AutoMLConfig` object. Additionally, specify your training data and the type of model, which is `regression` in this case.

In [11]:
from azureml.train.automl import AutoMLConfig

automl_config = AutoMLConfig(task='regression',
                             debug_log='automated_ml_errors.log',
                             training_data=x_train,
                             label_column_name="totalAmount",
                             **automl_settings)

Automated machine learning pre-processing steps (feature normalization, handling missing data, converting text to numeric, etc.) become part of the underlying model. When using the model for predictions, the same pre-processing steps applied during training are applied to your input data automatically.

### Train the automatic regression model

Create an experiment object in your workspace. An experiment acts as a container for your individual runs. Pass the defined `automl_config` object to the experiment, and set the output to `True` to view progress during the run. 

After starting the experiment, the output shown updates live as the experiment runs. For each iteration, you see the model type, the run duration, and the training accuracy. The field `BEST` tracks the best running training score based on your metric type.

In [12]:
from azureml.core.experiment import Experiment
experiment = Experiment(ws, "Tutorial-NYCTaxi-LIVE")
local_run = experiment.submit(automl_config, show_output=True)

No run_configuration provided, running on local with default configuration
Running in the active local environment.


Current status: DatasetEvaluation. Gathering dataset statistics.
Current status: FeaturesGeneration. Generating features for the dataset.
Current status: DatasetFeaturization. Beginning to fit featurizers and featurize the dataset.
Current status: DatasetFeaturizationCompleted. Completed fit featurizers and featurizing the dataset.
Current status: DatasetCrossValidationSplit. Generating individually featurized CV splits.

****************************************************************************************************
DATA GUARDRAILS: 

TYPE:         Missing feature values imputation
STATUS:       PASSED
DESCRIPTION:  No feature missing values were detected in the training data.
              Learn more about missing value imputation: https://aka.ms/AutomatedMLFeaturization

****************************************************************************************************

TYPE:         High cardinality feature detection
STATUS:       PASSED
DESCRIPTION:  Your inputs were analyzed

## Explore the results

Explore the results of automatic training with a [Jupyter widget](https://docs.microsoft.com/python/api/azureml-widgets/azureml.widgets?view=azure-ml-py). The widget allows you to see a graph and table of all individual run iterations, along with training accuracy metrics and metadata. Additionally, you can filter on different accuracy metrics than your primary metric with the dropdown selector.

In [13]:
from azureml.widgets import RunDetails
RunDetails(local_run).show()

_AutoMLWidget(widget_settings={'childWidgetDisplay': 'popup', 'send_telemetry': False, 'log_level': 'INFO', 's…

### Retrieve the best model

Select the best model from your iterations. The `get_output` function returns the best run and the fitted model for the last fit invocation. By using the overloads on `get_output`, you can retrieve the best run and fitted model for any logged metric or a particular iteration.

In [14]:
best_run, fitted_model = local_run.get_output()
print(best_run)
print(fitted_model)

Run(Experiment: Tutorial-NYCTaxi-LIVE,
Id: AutoML_a33261b2-50c2-411d-9b5c-b721eccd4fa7_30,
Type: None,
Status: Completed)
RegressionPipeline(pipeline=Pipeline(memory=None,
                                     steps=[('datatransformer',
                                             DataTransformer(enable_dnn=False, enable_feature_sweeping=True, feature_sweeping_config={}, feature_sweeping_timeout=86400, featurization_config=None, force_text_dnn=False, is_cross_validation=True, is_onnx_compatible=False, observer=None, task='regression', working_dir='/mnt/batch/ta...
    timeseries_param_dict=None,
    gpu_training_param_dict={'processing_unit_type': 'cpu'}
), random_state=0, reg_alpha=0, reg_lambda=0.8333333333333334, subsample=0.8, tree_method='auto'))], verbose=False))], weights=[0.13333333333333333, 0.06666666666666667, 0.06666666666666667, 0.4, 0.13333333333333333, 0.06666666666666667, 0.13333333333333333]))],
                                     verbose=False),
                   std

### Test the best model accuracy

Use the best model to run predictions on the test data set to predict taxi fares. The function `predict` uses the best model and predicts the values of y, **trip cost**, from the `x_test` data set. Print the first 10 predicted cost values from `y_predict`.

In [15]:
y_test = x_test.pop("totalAmount")

y_predict = fitted_model.predict(x_test)
print(y_predict[:10])

[13.15696369 10.05414941 14.614437    8.72629019 14.50762831  6.9117663
 19.94336619 12.88315871 11.009376   14.49779111]


Calculate the `root mean squared error` of the results. Convert the `y_test` dataframe to a list to compare to the predicted values. The function `mean_squared_error` takes two arrays of values and calculates the average squared error between them. Taking the square root of the result gives an error in the same units as the y variable, **cost**. It indicates roughly how far the taxi fare predictions are from the actual fares.

In [16]:
from sklearn.metrics import mean_squared_error
from math import sqrt

y_actual = y_test.values.flatten().tolist()
rmse = sqrt(mean_squared_error(y_actual, y_predict))
rmse

3.7136850574399456

Run the following code to calculate mean absolute percent error (MAPE) by using the full `y_actual` and `y_predict` data sets. This metric calculates an absolute difference between each predicted and actual value and sums all the differences. Then it expresses that sum as a percent of the total of the actual values.

In [17]:
sum_actuals = sum_errors = 0

for actual_val, predict_val in zip(y_actual, y_predict):
    abs_error = actual_val - predict_val
    if abs_error < 0:
        abs_error = abs_error * -1

    sum_errors = sum_errors + abs_error
    sum_actuals = sum_actuals + actual_val

mean_abs_percent_error = sum_errors / sum_actuals
print("Model MAPE:")
print(mean_abs_percent_error)
print()
print("Model Accuracy:")
print(1 - mean_abs_percent_error)

Model MAPE:
0.13593995627568936

Model Accuracy:
0.8640600437243107


From the two prediction accuracy metrics, you see that the model is fairly good at predicting taxi fares from the data set's features, typically within +- $4.00, and approximately 15% error. 

The traditional machine learning model development process is highly resource-intensive, and requires significant domain knowledge and time investment to run and compare the results of dozens of models. Using automated machine learning is a great way to rapidly test many different models for your scenario.

## Clean up resources

Do not complete this section if you plan on running other Azure Machine Learning service tutorials.

### Stop the notebook VM

If you used a cloud notebook server, stop the VM when you are not using it to reduce cost.

1. In your workspace, select **Compute**.
1. Select the **Notebook VMs** tab in the compute page.
1. From the list, select the VM.
1. Select **Stop**.
1. When you're ready to use the server again, select **Start**.

### Delete everything

If you don't plan to use the resources you created, delete them, so you don't incur any charges.

1. In the Azure portal, select **Resource groups** on the far left.
1. From the list, select the resource group you created.
1. Select **Delete resource group**.
1. Enter the resource group name. Then select **Delete**.

You can also keep the resource group but delete a single workspace. Display the workspace properties and select **Delete**.

## Next steps

In this automated machine learning tutorial, you did the following tasks:

> * Configured a workspace and prepared data for an experiment.
> * Trained by using an automated regression model locally with custom parameters.
> * Explored and reviewed training results.

[Deploy your model](https://docs.microsoft.com/azure/machine-learning/service/tutorial-deploy-models-with-aml) with Azure Machine Learning service.